exercises/exercise-14/database.ts

```diff
@@ -1,13 +1,112 @@
-export class Database<T> {
+import * as fs from 'fs';
+
+type JsonScalar = boolean | number | string;
+
+export type FieldOp = |
+    {$eq: JsonScalar} |
+    {$gt: JsonScalar} |
+    {$lt: JsonScalar} |
+    {$in: JsonScalar[]};
+
+export type Query<T extends object> = |
+    {$and: Query<T>[]} |
+    {$or: Query<T>[]} |
+    {$text: string} |
+    (
+        {[field in QueryableKeys<T>]?: FieldOp}
+    );
+
+function matchOp(op: FieldOp, v: JsonScalar) {
+    if ('$eq' in op) {
+        return v === op['$eq'];
+    } else if ('$gt' in op) {
+        return v > op['$gt'];
+    } else if ('$lt' in op) {
+        return v < op['$lt'];
+    } else if ('$in' in op) {
+        return op['$in'].includes(v);
+    }
+    throw new Error(`Unrecognized op: ${op}`);
+}
+
+type IndexedRecord<T extends object> = T & {
+    $index: {[word: string]: true};
+};
+
+type Unionize<T extends object> = {[k in keyof T]: {k: k, v: T[k]}}[keyof T];
+type QueryableKeys<T extends object> = Extract<Unionize<T>, {v: JsonScalar}>['k'];
+
+function matches<T extends object>(
+    q: Query<T>,
+    r: IndexedRecord<T>
+): boolean {
+    if ('$and' in q) {
+        return q.$and!.every(subq => matches(subq, r));
+    } else if ('$or' in q) {
+        return q.$or!.some(subq => matches(subq, r));
+    } else if ('$text' in q) {
+        const words = q.$text!.toLowerCase().split(' ');
+        return words.every(w => r.$index[w]);
+    }
+    return Object.entries(q).every(
+        ([k, v]) => matchOp(v as FieldOp, r[k as keyof T] as any)
+    );
+}
+
+export type Options<T> = {
+    sort?: {[k in keyof T]?: -1 | 1};
+    projection?: {[k in keyof T]?: 1}
+};
+
+function typedItems<T extends object>(o: T): [keyof T, T[keyof T]][] {
+    return Object.entries(o) as any;
+}
+
+function project<T extends object, K extends keyof T>(o: T, p: {[k in K]?: 1}): {[k in K]: T[k]} {
+    let out: Partial<T> = {};
+    for (const k in p) {
+        out[k] = o[k];
+    }
+    return out as any;
```

Comment:
how about return `Partial<T>` to prevent use `any`?
```
function project<T extends object, K extends keyof T>(o: T, p: {[k in K]?: 1}): Partial<T> {
  let out: Partial<T> = {};
  for (const k in p) {
    out[k] = o[k];
  }
  return out;
}
```